### ISSUE TITLE
Wagtail 4.1: modelcluster.fields.ParentalKey throws AttributeError: 'NoneType' object has no attribute '_meta'

### ISSUE THREAD

**Issue by MrCordeiro** [wagtail/wagtail#9583]
I'm been trying to update my blog to Wagtail 4.1 and I'm having issues with the following models.py setup:

```python
from modelcluster.models import ClusterableModel


@register_snippet
class BlogCategory(ClusterableModel):
    """
    A post category. Every blog post can only be in one category
    """

    name = models.CharField(max_length=255)

    panels = [ FieldPanel("name") ]

    class Meta:
        verbose_name = _("Category")
        verbose_name_plural = _("Categories")


class BlogPage(BasePage):
    """
    A Blog Page
    """
   # Other fields here

    category = ParentalKey(
        "BlogCategory",
        related_name="categories",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
    )
```

This code, for example, fails:

```
blog = BlogIndexPage(title="Blog", slug="blog")
post = BlogPage(title="post")
blog.add_child(instance=post)
```

The source of the error is[ this change in the `signal_handlers.py` module](https://github.com/wagtail/wagtail/commit/212b2e19a680410ac1ab3fe4f18cdcb318afaa67#diff-9e2fc90f09e08c4af86e25111803a306e91d17f5ad97333b51a2f2805226af2a) that was added on Wagtail 4.1.

More specifically, this signal:

```python
def update_reference_index_on_save(instance, **kwargs):
    # Don't populate reference index while loading fixtures as referenced objects may not be populated yet
    if kwargs.get("raw", False):
        return

    if getattr(reference_index_auto_update_disabled, "value", False):
        return

    # If the model is a child model, find the parent instance and index that instead
    while True:
        parental_keys = list(
            filter(
                lambda field: isinstance(field, ParentalKey),
                instance._meta.get_fields(),
            )
        )
        if not parental_keys:
            break

        instance = getattr(instance, parental_keys[0].name)

    if ReferenceIndex.model_is_indexable(type(instance)):
        with transaction.atomic():
            ReferenceIndex.create_or_update_for_object(instance)
```

Since an instance with the `parental_keys` var is populated even when the field is `null`, the `instance` variable becomes null and we get an error. 

### Technical details

-   Python version: Python 3.10.8
-   Django version: Django 4.1.6
-   Wagtail version: Wagtail 4.1

**Comment by gasman** [comment]
Hi @MrCordeiro - thanks for the report.
I should note that the use of ParentalKey isn't really correct here - it implies that any BlogPage record is _logically part of_ the BlogCategory object. In particular, I suspect that whenever you make an edit to the BlogCategory object through the Wagtail admin, it'll end up saving a copy of every linked BlogPage as part of that object's revision history. You should use a ForeignKey for this instead.

However, it's still not ideal for the signal handler to fail in this situation, so I'll leave this issue open to be addressed in the next 4.1 bugfix release.

**Comment by MrCordeiro** [comment]
Ah thanks for the feedback! I didn't know about that!Am 02.11.2022 8:50 nachm. schrieb Matt Westcott ***@***.***>:
Hi @MrCordeiro - thanks for the report.
I should note that the use of ParentalKey isn't really correct here - it implies that any BlogPage record is logically part of the BlogCategory object. In particular, I suspect that whenever you make an edit to the BlogCategory object through the Wagtail admin, it'll end up saving a copy of every linked BlogPage as part of that object's revision history. You should use a ForeignKey for this instead.
However, it's still not ideal for the signal handler to fail in this situation, so I'll leave this issue open to be addressed in the next 4.1 bugfix release.
